docs/RFCS/20200414_scheduled_jobs.md

```diff
@@ -0,0 +1,424 @@
+This hook will lookup the job, and invoke `ScheduledJobExecutor`
+to notify it of the job completion.
+
+
+#### Polling
+Our initial MVP version will use a simple polling mechanism where each node picks 
+the next runnable job, and attempts to execute it.
+
+Inside the transaction, we select the next eligible job to run.
+
+```sql
+SELECT 
+  C.job, C.schedule, C.schedule_details, C.user,
+  (SELECT count(*) 
+   FROM system.jobs J
+   WHERE J.sched_id = C.sched_id AND job_status NOT IN ('failed', 'succeeded')
+  ) AS num_running,
+FROM system.scheduled_jobs C
+WHERE next_run < current_timestamp()
+```
+
+If `num_running > 0` and policy setting `skip_wait == false`, then we need to
+wait for the previous invocation to complete.  To do this, we simply advance
+the `next_run` time by some amount and commit transaction.
+
+Once job execution policy constraints have been satisfied, 
+we can compute the next time this job is supposed to run.
+If the schedule is set, then compute the next_run based on the schedule.
+If not, this is a one-off job, set `next_run` to `NULL` (and update
+`change_info` to indicate that the schedule completed).
+
+Finally, execute the statement and commit the transaction.
+
+We have multiple venues on improving scalability and performance in the future releases,
+driven by the customer needs (Cockroach Cloud, etc): 
+  * Nodes may pick a random runnable job (instead of the oldest), to reduce contention
+  * If polling becomes problematic (scale), we can explore using another mechanism 
+  (leasing/locks) to coordinate which node has exclusive dispatch responsibility,
+   and then allow that node to load the schedule into a more efficient structure 
+   like a timer ring that is populated via polling or a rangefeed.
+  * We can add a coordinator job responsible for finding lots of work to do and 
+    spawning these jobs
+  * Loading some (or all) scheduled state in memory to avoid rescans
+
+#### Pause/Resume
+Pausing the schedule involves setting `next_run` to NULL (and updating
+change info appropriately, e.g. reason="paused by an operator")
+
+Un-pausing the schedule involves computing the `next_run` time based on the `schedule`
+(and recording the change info)
+
+#### One-Off Jobs
+It is possible to use scheduled jobs to execute a one-off job.  A one-off job is represented
+by having an entry in the `system.scheduled_jobs` table that has a `next_run` set, but does not
+have a `schedule_expr`.
+
+## Monitoring 
+
+Since scheduled jobs is responsible for running unattended jobs, it is import that
+we provide visibility into the health of those jobs.
+
+The full details of the monitoring integration are outside the scope
+of this document, however, at the very least, we will export the following
+metrics:
+   * number of paused scheduled jobs
+   * number of jobs started
+   
+## CLI / Admin UI
+
+Interaction with the scheduled job system will require CLI and UI support, as well as changes
```

Comment:
Does this imply that schedules can only be created through a CLI command? Or will there be a SQL interface to the scheduled job system?

Reply:
I believe this was meant to mean _SQL_ CLI.